framework/src/main/java/org/checkerframework/common/wholeprograminference/scenelib/ASceneWrapper.java

```diff
@@ -0,0 +1,250 @@
+     * @param classSymbol the element representing the class, used for adding data to the
+     *     AClassWrapper returned by this method. If it is null, an AClassWrapper is looked up or
+     *     created, but the other information stored by the AClassWrapper is not updated.
+     * @return an AClassWrapper representing that class
+     */
+    public AClassWrapper vivifyClass(
+            @BinaryName String className, @Nullable ClassSymbol classSymbol) {
+        AClassWrapper wrapper;
+        if (classes.containsKey(className)) {
+            wrapper = classes.get(className);
+        } else {
+            AClass aClass = theScene.classes.getVivify(className);
+            wrapper = new AClassWrapper(aClass);
+            classes.put(className, wrapper);
+        }
+
+        // updateClassData must be called on both paths (cache hit and cache miss) because the
+        // second parameter could have been null when the first miss occurred.
+        // Different visit methods in CFAbstractTransfer call WPI in different ways.  Only some
+        // provide the metadata, and the visit order isn't known ahead of time.
+        // Since it is not used until the end of WPI, it being unavailable during WPI is not a
+        // problem.
+        if (classSymbol != null) {
+            updateClassData(wrapper, classSymbol);
+        }
+        return wrapper;
+    }
+
+    /**
+     * Updates the metadata stored in AClassWrapper for the given class.
+     *
+     * @param aClassWrapper the class representation in which the metadata is to be updated
+     * @param classSymbol the class for which to update metadata
+     */
+    private void updateClassData(AClassWrapper aClassWrapper, ClassSymbol classSymbol) {
+        if (classSymbol.isEnum()) {
+            if (!aClassWrapper.isEnum(classSymbol.getSimpleName().toString())) {
+                List<VariableElement> enumConstants = new ArrayList<>();
+                for (Element e : ((TypeElement) classSymbol).getEnclosedElements()) {
+                    if (e.getKind() == ElementKind.ENUM_CONSTANT) {
+                        enumConstants.add((VariableElement) e);
+                    }
+                }
+                aClassWrapper.setEnumConstants(enumConstants);
+            }
+        }
+
+        ClassSymbol outerClass = classSymbol;
+        ClassSymbol previous = classSymbol;
+        do {
+            if (outerClass.isEnum()) {
+                aClassWrapper.markAsEnum(outerClass.getSimpleName().toString());
+            }
+            Element element = classSymbol.getEnclosingElement();
+            if (element == null || element.getKind() == ElementKind.PACKAGE) {
+                break;
+            }
+            TypeElement t = ElementUtils.enclosingClass(element);
+            previous = outerClass;
+            outerClass = (ClassSymbol) t;
+        } while (outerClass != null && !previous.equals(outerClass));
```

Comment:
I added a note to explain that the test is necessary to prevent an infinite loop. I'm not sure what causes the infinite loop; the list of test classes that cause it to trigger has no unifying characteristic, as far as I can tell. I generated the list by adding this code at the end of the loop:

```
            if (previous.equals(outerClass)) {
                System.out.println(outerClass);
            }
```

If you have any idea why it's necessary, I'd love to hear it. Here's the list:

```    
    EnumWithInnerClass
    EnumWithInnerClass
    EnumWithInnerClass
    ExpectedErrors
    ExpectedErrors
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    EnumTest
    DeviceTypeTest
    DeviceTypeTest
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    UsesAnonymous
    <anonymous NamedInnerClassInAnonymous$1>
    MethodReturnTest
    MethodReturnTest
    MethodReturnTest
    MethodReturnTest
    MethodReturnTest

```